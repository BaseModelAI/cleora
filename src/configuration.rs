/// Pipeline configuration
#[derive(Debug)]
pub struct Configuration {
    /// Produce or not entity counter to the output file
    pub produce_entity_occurrence_count: bool,

    /// Dimension of the embedding
    pub embeddings_dimension: u16,

    /// Maximum number of iteration for training
    pub max_number_of_iteration: u8,

    /// Prepend field name to entity in the output file. It differentiates entities with the same
    /// name from different columns
    pub prepend_field: bool,

    /// After how many lines we log the progress
    pub log_every_n: u32,

    /// Calculate embeddings in memory or with memory-mapped files. If we don't have enough
    /// RAM we can support training with mmap files
    pub in_memory_embedding_calculation: bool,

    /// Path to the input file
    pub input: String,

    /// Output directory for files with embeddings
    pub output_dir: Option<String>,

    /// Name of the relation, for output filename generation
    pub relation_name: String,

    /// Columns configuration
    pub columns: Vec<Column>,
}

/// Column configuration
#[derive(Debug, Default)]
pub struct Column {
    /// Name, header of the column
    pub name: String,

    /// The field is virtual - it is considered during embedding process, no entity is written for the column
    pub transient: bool,

    /// The field is composite, containing multiple entity identifiers separated by space
    pub complex: bool,

    /// The field is reflexive, which means that it interacts with itself, additional output file is written for every such field
    pub reflexive: bool,

    /// The field is ignored, no output file is written for the field
    pub ignored: bool,
}

impl Configuration {
    /// Create default configuration with specified input file path and columns.
    pub fn default(input: String, columns: Vec<Column>) -> Configuration {
        Configuration {
            produce_entity_occurrence_count: true,
            embeddings_dimension: 128,
            max_number_of_iteration: 4,
            prepend_field: true,
            log_every_n: 1000,
            in_memory_embedding_calculation: true,
            input,
            output_dir: None,
            relation_name: String::from("emb"),
            columns,
        }
    }

    /// Filter out ignored columns. Entities from such columns are omitted.
    pub fn not_ignored_columns(&self) -> Vec<&Column> {
        self.columns.iter().filter(|&c| !c.ignored).collect()
    }
}

/// Extract columns config based on raw strings.
pub fn extract_fields(cols: Vec<&str>) -> Result<Vec<Column>, String> {
    let mut columns: Vec<Column> = Vec::new();

    for col in cols {
        let parts: Vec<&str> = col.split("::").collect();

        let column_name: &str;
        let mut transient = false;
        let mut complex = false;
        let mut reflexive = false;
        let mut ignored = false;

        let parts_len = parts.len();
        if parts_len > 1 {
            column_name = *parts.last().unwrap();
            let column_name_idx = parts_len - 1;
            for &part in &parts[..column_name_idx] {
                if part.eq_ignore_ascii_case("transient") {
                    transient = true;
                } else if part.eq_ignore_ascii_case("complex") {
                    complex = true;
                } else if part.eq_ignore_ascii_case("reflexive") {
                    reflexive = true;
                } else if part.eq_ignore_ascii_case("ignore") {
                    ignored = true;
                } else {
                    let message = format!("Unrecognized column field modifier: {}", part);
                    return Err(message);
                }
            }
        } else {
            column_name = col;
        }
        let column = Column {
            name: column_name.to_string(),
            transient,
            complex,
            reflexive,
            ignored,
        };
        columns.push(column);
    }
    Ok(columns)
}

/// Validate column modifiers.
pub fn validate_fields(cols: Vec<Column>) -> Result<Vec<Column>, String> {
    for col in &cols {
        // transient::reflexive - this would generate no output
        // transient::reflexive::complex - this would generate no output
        if col.reflexive && col.transient {
            let message = format!("A field cannot be REFLEXIVE and simultaneously TRANSIENT. It does not make sense: {}", col.name);
            return Err(message);
        }
        if col.reflexive && !col.complex {
            let message = format!(
                "A field cannot be REFLEXIVE but NOT COMPLEX. It does not make sense: {}",
                col.name
            );
            return Err(message);
        }
    }
    Ok(cols)
}
