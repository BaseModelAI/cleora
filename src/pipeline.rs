use std::fs::File;
use std::io::{BufRead, BufReader};

use crate::configuration::{Column, Configuration, FileType, OutputFormat};
use crate::embedding::{calculate_embeddings, calculate_embeddings_mmap};
use crate::entity::{EntityProcessor, SMALL_VECTOR_SIZE};
use crate::persistence::embedding::{EmbeddingPersistor, NpyPersistor, TextFileVectorPersistor};
use crate::persistence::entity::InMemoryEntityMappingPersistor;
use crate::sparse_matrix::{create_sparse_matrices, SparseMatrix};
use bus::Bus;
use log::{error, info};
use simdjson_rust::dom;
use smallvec::{smallvec, SmallVec};
use std::sync::Arc;
use std::thread;

/// Create SparseMatrix'es based on columns config. Every SparseMatrix operates in separate
/// thread. EntityProcessor reads data in main thread and broadcast cartesian products
/// to SparseMatrix'es.
pub fn build_graphs(
    config: &Configuration,
    in_memory_entity_mapping_persistor: Arc<InMemoryEntityMappingPersistor>,
) -> Vec<SparseMatrix> {
    let sparse_matrices = create_sparse_matrices(&config.columns);
    dbg!(&sparse_matrices);

    let mut bus: Bus<SmallVec<[u64; SMALL_VECTOR_SIZE]>> = Bus::new(128);
    let mut sparse_matrix_threads = Vec::new();
    for mut sparse_matrix in sparse_matrices {
        let rx = bus.add_rx();
        let handle = thread::spawn(move || {
            for received in rx {
                if received[0] != 0 {
                    sparse_matrix.handle_pair(&received);
                } else {
                    sparse_matrix.finish();
                    break;
                }
            }
            sparse_matrix
        });
        sparse_matrix_threads.push(handle);
    }

    let mut entity_processor =
        EntityProcessor::new(&config, in_memory_entity_mapping_persistor, |hashes| {
            bus.broadcast(hashes);
        });

    match &config.file_type {
        FileType::JSON => {
            let mut parser = dom::Parser::default();
            read_file(config, |line| {
                let row = parse_json_line(line, &mut parser, &config.columns);
                entity_processor.process_row(&row);
            });
        }
        FileType::TSV => {
            read_file(config, |line| {
                let row = parse_tsv_line(line);
                entity_processor.process_row(&row);
            });
        }
    }

    entity_processor.finish();

    let mut sparse_matrices = vec![];
    for join_handle in sparse_matrix_threads {
        let sparse_matrix = join_handle
            .join()
            .expect("Couldn't join on the associated thread");
        sparse_matrices.push(sparse_matrix);
    }

    sparse_matrices
}

/// Read file line by line. Pass every valid line to handler for parsing.
fn read_file<F>(config: &Configuration, mut line_handler: F)
where
    F: FnMut(&str),
{
    let input_file = File::open(&config.input).expect("Can't open file");
    let mut buffered = BufReader::new(input_file);

    let mut line_number = 1u64;
    let mut line = String::new();
    loop {
        match buffered.read_line(&mut line) {
            Ok(bytes_read) => {
                // EOF
                if bytes_read == 0 {
                    break;
                }

                line_handler(&line);
            }
            Err(err) => {
                error!("Can't read line number: {}. Error: {}.", line_number, err);
            }
        };

        // clear to reuse the buffer
        line.clear();

        if line_number % config.log_every_n as u64 == 0 {
            info!("Number of lines processed: {}", line_number);
        }

        line_number += 1;
    }
}

/// Parse a line of JSON and read its columns into a vector for processing.
fn parse_json_line(
    line: &str,
    parser: &mut dom::Parser,
    columns: &[Column],
) -> Vec<SmallVec<[String; SMALL_VECTOR_SIZE]>> {
    let parsed = parser.parse(&line).unwrap();
    columns
        .iter()
        .map(|c| {
            if !c.complex {
                let elem = parsed.at_key(&c.name).unwrap();
                let value = match elem.get_type() {
                    dom::element::ElementType::String => elem.get_string().unwrap(),
                    _ => elem.minify(),
                };
                smallvec![value]
            } else {
                parsed
                    .at_key(&c.name)
                    .unwrap()
                    .get_array()
                    .expect("Values for complex columns must be arrays")
                    .into_iter()
                    .map(|v| match v.get_type() {
                        dom::element::ElementType::String => v.get_string().unwrap(),
                        _ => v.minify(),
                    })
                    .collect()
            }
        })
        .collect()
}

/// Parse a line of TSV and read its columns into a vector for processing.
fn parse_tsv_line(line: &str) -> Vec<SmallVec<[&str; SMALL_VECTOR_SIZE]>> {
    let values = line.trim().split('\t');
    values.map(|c| c.split(' ').collect()).collect()
}

/// Train SparseMatrix'es (graphs) in separated threads.
pub fn train(
    config: Configuration,
    in_memory_entity_mapping_persistor: Arc<InMemoryEntityMappingPersistor>,
    sparse_matrices: Vec<SparseMatrix>,
) {
    let config = Arc::new(config);
    let mut embedding_threads = Vec::new();
    for sparse_matrix in sparse_matrices {
        let sparse_matrix = Arc::new(sparse_matrix);
        let config = config.clone();
        let in_memory_entity_mapping_persistor = in_memory_entity_mapping_persistor.clone();
        let handle = thread::spawn(move || {
            let directory = match config.output_dir.as_ref() {
                Some(out) => format!("{}/", out.clone()),
                None => String::from(""),
            };
            let ofp = format!(
                "{}{}__{}__{}.out",
                directory,
                config.relation_name,
                sparse_matrix.col_a_name.as_str(),
                sparse_matrix.col_b_name.as_str()
            );

            let mut persistor: Box<dyn EmbeddingPersistor> = match &config.output_format {
                OutputFormat::TextFile => Box::new(TextFileVectorPersistor::new(
                    ofp,
                    config.produce_entity_occurrence_count,
                )),
                OutputFormat::Numpy => Box::new(NpyPersistor::new(
                    ofp,
                    config.produce_entity_occurrence_count,
                )),
            };
            if config.in_memory_embedding_calculation {
                calculate_embeddings(
                    config.clone(),
                    sparse_matrix.clone(),
                    in_memory_entity_mapping_persistor,
                    persistor.as_mut(),
                );
            } else {
                calculate_embeddings_mmap(
                    config.clone(),
                    sparse_matrix.clone(),
                    in_memory_entity_mapping_persistor,
                    persistor.as_mut(),
                );
            }
        });
        embedding_threads.push(handle);
    }

    for join_handle in embedding_threads {
        let _ = join_handle
            .join()
            .expect("Couldn't join on the associated thread");
    }
}
