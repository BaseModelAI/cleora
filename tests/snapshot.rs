use cleora::configuration::{Column, Configuration};
use cleora::embedding::calculate_embeddings;
use cleora::persistence::embedding::EmbeddingPersistor;
use cleora::persistence::entity::InMemoryEntityMappingPersistor;
use cleora::pipeline::build_graphs;
use insta::assert_debug_snapshot;
use std::sync::Arc;

/// This test performs work for sample case and saves snapshot file.
/// Snapshot testing takes advantage of deterministic character of Cleora.
/// Any discrepancies between original snapshot results and current ones can be then
/// reviewed along with the code which introduced discrepancy.
///
/// Differing snapshot has to be renamed by removing .new from the name.
/// For more information, please review https://crates.io/crates/insta
///
/// Code executed performs roughly the same work as:
/// ./cleora -i files/samples/edgelist_1.tsv --columns="complex::reflexive::a b complex::c"
/// -d 128 -n 4 --relation-name=R1 -p 0
#[test]
fn test_build_graphs_and_create_embedding() {
    let columns = vec![
        Column {
            name: "a".to_string(),
            complex: true,
            reflexive: true,
            ..Column::default()
        },
        Column {
            name: "b".to_string(),
            ..Column::default()
        },
        Column {
            name: "c".to_string(),
            complex: true,
            ..Column::default()
        },
    ];

    let config = Configuration {
        produce_entity_occurrence_count: true,
        embeddings_dimension: 128,
        max_number_of_iteration: 4,
        prepend_field: false,
        log_every_n: 10000,
        in_memory_embedding_calculation: true,
        input: "files/samples/edgelist_1.tsv".to_string(),
        output_dir: None,
        relation_name: "r1".to_string(),
        columns,
    };
    let in_memory_entity_mapping_persistor = InMemoryEntityMappingPersistor::new();
    let in_memory_entity_mapping_persistor = Arc::new(in_memory_entity_mapping_persistor);

    // build sparse matrices
    let mut sparse_matrices = build_graphs(&config, in_memory_entity_mapping_persistor.clone());
    assert_debug_snapshot!("sparse_matrices", sparse_matrices);

    let mut in_memory_embedding_persistor = InMemoryEmbeddingPersistor::default();

    // calculate embeddings for ONE sparse matrix
    calculate_embeddings(
        &mut sparse_matrices[0],
        config.max_number_of_iteration,
        in_memory_entity_mapping_persistor,
        &mut in_memory_embedding_persistor,
    );

    assert_debug_snapshot!("embeddings", in_memory_embedding_persistor);
}

#[derive(Debug, Default)]
pub struct InMemoryEmbeddingPersistor {
    entity_count: u32,
    dimenstion: u16,
    entities: Vec<InMemoryEntity>,
}

#[derive(Debug)]
pub struct InMemoryEntity {
    entity: String,
    occur_count: u32,
    vector: Vec<f32>,
}

impl EmbeddingPersistor for InMemoryEmbeddingPersistor {
    fn put_metadata(&mut self, entity_count: u32, dimension: u16) {
        self.entity_count = entity_count;
        self.dimenstion = dimension;
    }
    fn put_data(&mut self, entity: String, occur_count: u32, vector: Vec<f32>) {
        self.entities.push(InMemoryEntity {
            entity,
            occur_count,
            vector,
        });
    }
    fn finish(&mut self) {}
}
