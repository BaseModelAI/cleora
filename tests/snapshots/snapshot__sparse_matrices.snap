---
source: tests/snapshot.rs
expression: sparse_matrices
---
[
    SparseMatrix {
        col_a_id: 0,
        col_a_name: "a",
        col_b_id: 3,
        col_b_name: "a",
        dimension: 128,
        sparse_matrix_persistor: InMemorySparseMatrixPersistor {
            entity_count: 6,
            edge_count: 77,
            hash_2_id: {
                0: 0,
                13238837134403314746: 3,
                14710143206259780207: 4,
                2596444481629440690: 5,
                5785750101991912937: 1,
                18279053511379946123: 2,
            },
            id_2_hash: {
                0: 0,
                5: 2596444481629440690,
                2: 18279053511379946123,
                4: 14710143206259780207,
                1: 5785750101991912937,
                3: 13238837134403314746,
            },
            hash_2_count: {
                0: 46,
                13238837134403314746: 28,
                14710143206259780207: 12,
                2596444481629440690: 16,
                5785750101991912937: 34,
                18279053511379946123: 18,
            },
            row_sum: [
                1.8333336,
                1.1666666,
                0.6666666,
                1.1666667,
                0.66666675,
                0.5,
            ],
            pair_index: {
                0: 0,
                35: 25,
                26: 19,
                23: 15,
                20: 12,
                8: 8,
                5: 6,
                2: 2,
                31: 20,
                28: 21,
                25: 23,
                19: 14,
                16: 11,
                13: 18,
                10: 17,
                7: 7,
                4: 3,
                1: 1,
                33: 22,
                30: 24,
                24: 16,
                15: 13,
                12: 10,
                9: 9,
                6: 4,
                3: 5,
            },
            entries: [
                Entry {
                    row: 0,
                    col: 0,
                    value: 0.5694444,
                },
                Entry {
                    row: 0,
                    col: 1,
                    value: 0.34722224,
                },
                Entry {
                    row: 1,
                    col: 0,
                    value: 0.34722224,
                },
                Entry {
                    row: 0,
                    col: 2,
                    value: 0.22222224,
                },
                Entry {
                    row: 2,
                    col: 0,
                    value: 0.22222224,
                },
                Entry {
                    row: 1,
                    col: 1,
                    value: 0.34722224,
                },
                Entry {
                    row: 1,
                    col: 2,
                    value: 0.22222224,
                },
                Entry {
                    row: 2,
                    col: 1,
                    value: 0.22222224,
                },
                Entry {
                    row: 2,
                    col: 2,
                    value: 0.22222224,
                },
                Entry {
                    row: 0,
                    col: 3,
                    value: 0.3472222,
                },
                Entry {
                    row: 3,
                    col: 0,
                    value: 0.3472222,
                },
                Entry {
                    row: 0,
                    col: 4,
                    value: 0.22222222,
                },
                Entry {
                    row: 4,
                    col: 0,
                    value: 0.22222222,
                },
                Entry {
                    row: 3,
                    col: 3,
                    value: 0.3472222,
                },
                Entry {
                    row: 3,
                    col: 4,
                    value: 0.22222222,
                },
                Entry {
                    row: 4,
                    col: 3,
                    value: 0.22222222,
                },
                Entry {
                    row: 4,
                    col: 4,
                    value: 0.22222222,
                },
                Entry {
                    row: 1,
                    col: 3,
                    value: 0.125,
                },
                Entry {
                    row: 3,
                    col: 1,
                    value: 0.125,
                },
                Entry {
                    row: 1,
                    col: 5,
                    value: 0.125,
                },
                Entry {
                    row: 5,
                    col: 1,
                    value: 0.125,
                },
                Entry {
                    row: 3,
                    col: 5,
                    value: 0.125,
                },
                Entry {
                    row: 5,
                    col: 3,
                    value: 0.125,
                },
                Entry {
                    row: 0,
                    col: 5,
                    value: 0.125,
                },
                Entry {
                    row: 5,
                    col: 0,
                    value: 0.125,
                },
                Entry {
                    row: 5,
                    col: 5,
                    value: 0.125,
                },
            ],
        },
    },
    SparseMatrix {
        col_a_id: 0,
        col_a_name: "a",
        col_b_id: 1,
        col_b_name: "b",
        dimension: 128,
        sparse_matrix_persistor: InMemorySparseMatrixPersistor {
            entity_count: 8,
            edge_count: 77,
            hash_2_id: {
                0: 0,
                8103484332547480218: 1,
                13238837134403314746: 5,
                18279053511379946123: 3,
                2596444481629440690: 7,
                10392296954106637551: 4,
                14710143206259780207: 6,
                5785750101991912937: 2,
            },
            id_2_hash: {
                0: 0,
                7: 2596444481629440690,
                4: 10392296954106637551,
                1: 8103484332547480218,
                5: 13238837134403314746,
                2: 5785750101991912937,
                6: 14710143206259780207,
                3: 18279053511379946123,
            },
            hash_2_count: {
                0: 23,
                8103484332547480218: 59,
                13238837134403314746: 14,
                18279053511379946123: 9,
                2596444481629440690: 8,
                10392296954106637551: 18,
                14710143206259780207: 6,
                5785750101991912937: 17,
            },
            row_sum: [
                0.9166668,
                1.9999998,
                0.5833334,
                0.33333337,
                1.0000002,
                0.5833334,
                0.33333334,
                0.25,
            ],
            pair_index: {
                29: 9,
                26: 13,
                20: 7,
                46: 10,
                40: 11,
                5: 3,
                2: 1,
                34: 8,
                31: 12,
                57: 14,
                16: 6,
                13: 4,
                10: 5,
                7: 2,
                1: 0,
                50: 15,
            },
            entries: [
                Entry {
                    row: 0,
                    col: 1,
                    value: 0.5833334,
                },
                Entry {
                    row: 1,
                    col: 0,
                    value: 0.5833334,
                },
                Entry {
                    row: 2,
                    col: 1,
                    value: 0.5833334,
                },
                Entry {
                    row: 1,
                    col: 2,
                    value: 0.5833334,
                },
                Entry {
                    row: 3,
                    col: 1,
                    value: 0.33333337,
                },
                Entry {
                    row: 1,
                    col: 3,
                    value: 0.33333337,
                },
                Entry {
                    row: 0,
                    col: 4,
                    value: 0.33333334,
                },
                Entry {
                    row: 4,
                    col: 0,
                    value: 0.33333334,
                },
                Entry {
                    row: 5,
                    col: 4,
                    value: 0.33333334,
                },
                Entry {
                    row: 4,
                    col: 5,
                    value: 0.33333334,
                },
                Entry {
                    row: 6,
                    col: 4,
                    value: 0.33333334,
                },
                Entry {
                    row: 4,
                    col: 6,
                    value: 0.33333334,
                },
                Entry {
                    row: 5,
                    col: 1,
                    value: 0.25,
                },
                Entry {
                    row: 1,
                    col: 5,
                    value: 0.25,
                },
                Entry {
                    row: 7,
                    col: 1,
                    value: 0.25,
                },
                Entry {
                    row: 1,
                    col: 7,
                    value: 0.25,
                },
            ],
        },
    },
    SparseMatrix {
        col_a_id: 0,
        col_a_name: "a",
        col_b_id: 2,
        col_b_name: "c",
        dimension: 128,
        sparse_matrix_persistor: InMemorySparseMatrixPersistor {
            entity_count: 9,
            edge_count: 77,
            hash_2_id: {
                0: 0,
                15837096338808335664: 3,
                13238837134403314746: 6,
                15555829607237049207: 2,
                18279053511379946123: 5,
                16291251679096942939: 1,
                2596444481629440690: 8,
                14710143206259780207: 7,
                5785750101991912937: 4,
            },
            id_2_hash: {
                0: 0,
                7: 14710143206259780207,
                4: 5785750101991912937,
                1: 16291251679096942939,
                8: 2596444481629440690,
                5: 18279053511379946123,
                2: 15555829607237049207,
                6: 13238837134403314746,
                3: 15837096338808335664,
            },
            hash_2_count: {
                0: 23,
                15837096338808335664: 18,
                13238837134403314746: 14,
                15555829607237049207: 25,
                18279053511379946123: 9,
                16291251679096942939: 34,
                2596444481629440690: 8,
                14710143206259780207: 6,
                5785750101991912937: 17,
            },
            row_sum: [
                0.9166668,
                1.3333335,
                0.8333334,
                0.83333355,
                0.5833334,
                0.33333337,
                0.5833334,
                0.33333334,
                0.25,
            ],
            pair_index: {
                52: 23,
                43: 20,
                37: 21,
                31: 12,
                28: 17,
                22: 8,
                19: 11,
                74: 30,
                4: 2,
                1: 0,
                65: 29,
                59: 22,
                50: 25,
                44: 26,
                38: 27,
                32: 14,
                26: 13,
                23: 10,
                17: 7,
                2: 1,
                66: 31,
                57: 24,
                45: 18,
                39: 19,
                33: 16,
                27: 15,
                21: 6,
                18: 9,
                12: 5,
                9: 4,
                6: 3,
                73: 28,
            },
            entries: [
                Entry {
                    row: 0,
                    col: 1,
                    value: 0.4027778,
                },
                Entry {
                    row: 1,
                    col: 0,
                    value: 0.4027778,
                },
                Entry {
                    row: 0,
                    col: 2,
                    value: 0.2361111,
                },
                Entry {
                    row: 2,
                    col: 0,
                    value: 0.2361111,
                },
                Entry {
                    row: 0,
                    col: 3,
                    value: 0.2777778,
                },
                Entry {
                    row: 3,
                    col: 0,
                    value: 0.2777778,
                },
                Entry {
                    row: 4,
                    col: 1,
                    value: 0.2361111,
                },
                Entry {
                    row: 1,
                    col: 4,
                    value: 0.2361111,
                },
                Entry {
                    row: 4,
                    col: 2,
                    value: 0.2361111,
                },
                Entry {
                    row: 2,
                    col: 4,
                    value: 0.2361111,
                },
                Entry {
                    row: 4,
                    col: 3,
                    value: 0.11111111,
                },
                Entry {
                    row: 3,
                    col: 4,
                    value: 0.11111111,
                },
                Entry {
                    row: 5,
                    col: 1,
                    value: 0.11111111,
                },
                Entry {
                    row: 1,
                    col: 5,
                    value: 0.11111111,
                },
                Entry {
                    row: 5,
                    col: 2,
                    value: 0.11111111,
                },
                Entry {
                    row: 2,
                    col: 5,
                    value: 0.11111111,
                },
                Entry {
                    row: 5,
                    col: 3,
                    value: 0.11111111,
                },
                Entry {
                    row: 3,
                    col: 5,
                    value: 0.11111111,
                },
                Entry {
                    row: 6,
                    col: 3,
                    value: 0.16666667,
                },
                Entry {
                    row: 3,
                    col: 6,
                    value: 0.16666667,
                },
                Entry {
                    row: 6,
                    col: 1,
                    value: 0.2916667,
                },
                Entry {
                    row: 1,
                    col: 6,
                    value: 0.2916667,
                },
                Entry {
                    row: 7,
                    col: 3,
                    value: 0.16666667,
                },
                Entry {
                    row: 3,
                    col: 7,
                    value: 0.16666667,
                },
                Entry {
                    row: 7,
                    col: 1,
                    value: 0.16666667,
                },
                Entry {
                    row: 1,
                    col: 7,
                    value: 0.16666667,
                },
                Entry {
                    row: 6,
                    col: 2,
                    value: 0.125,
                },
                Entry {
                    row: 2,
                    col: 6,
                    value: 0.125,
                },
                Entry {
                    row: 8,
                    col: 1,
                    value: 0.125,
                },
                Entry {
                    row: 1,
                    col: 8,
                    value: 0.125,
                },
                Entry {
                    row: 8,
                    col: 2,
                    value: 0.125,
                },
                Entry {
                    row: 2,
                    col: 8,
                    value: 0.125,
                },
            ],
        },
    },
    SparseMatrix {
        col_a_id: 1,
        col_a_name: "b",
        col_b_id: 2,
        col_b_name: "c",
        dimension: 128,
        sparse_matrix_persistor: InMemorySparseMatrixPersistor {
            entity_count: 5,
            edge_count: 77,
            hash_2_id: {
                15837096338808335664: 3,
                8103484332547480218: 0,
                15555829607237049207: 2,
                10392296954106637551: 4,
                16291251679096942939: 1,
            },
            id_2_hash: {
                0: 8103484332547480218,
                2: 15555829607237049207,
                4: 10392296954106637551,
                1: 16291251679096942939,
                3: 15837096338808335664,
            },
            hash_2_count: {
                15837096338808335664: 18,
                8103484332547480218: 59,
                15555829607237049207: 25,
                10392296954106637551: 18,
                16291251679096942939: 34,
            },
            row_sum: [
                1.9999998,
                1.3333335,
                0.8333334,
                0.83333355,
                1.0000002,
            ],
            pair_index: {
                23: 6,
                4: 2,
                1: 0,
                17: 9,
                21: 8,
                2: 1,
                12: 5,
                9: 4,
                6: 3,
                19: 7,
            },
            entries: [
                Entry {
                    row: 0,
                    col: 1,
                    value: 0.8333334,
                },
                Entry {
                    row: 1,
                    col: 0,
                    value: 0.8333334,
                },
                Entry {
                    row: 0,
                    col: 2,
                    value: 0.8333334,
                },
                Entry {
                    row: 2,
                    col: 0,
                    value: 0.8333334,
                },
                Entry {
                    row: 0,
                    col: 3,
                    value: 0.33333337,
                },
                Entry {
                    row: 3,
                    col: 0,
                    value: 0.33333337,
                },
                Entry {
                    row: 4,
                    col: 3,
                    value: 0.5,
                },
                Entry {
                    row: 3,
                    col: 4,
                    value: 0.5,
                },
                Entry {
                    row: 4,
                    col: 1,
                    value: 0.5,
                },
                Entry {
                    row: 1,
                    col: 4,
                    value: 0.5,
                },
            ],
        },
    },
]
